### PR TITLE
httpauth: basic challenge creation and verification functions

### PR DESCRIPTION
**Pull request by cHuberCoffee**
+ create and encode HTTP Basic requests.
+ verify HTTP basic authentication header
+ test cases for httpauth_basic verification and request encoding
+ httpauth_basic is tested against wget, curl and an chromium based browser (Brave Browser)

### CODE REVIEW

File: src/httpauth/basic.c [hunk 1]
```diff
@@ -123,3 +123,119 @@ int httpauth_basic_encode(const struct httpauth_basic *basic, struct mbuf *mb)
...
+			req->realm, req->charset);
+	else
+		return re_hprintf(pf, "Basic realm=\"%s\"", req->realm);
+}
+
+
+/**
+ * Verify received credentials
+ *
+ * @param user   user name (may be an UTF-8 string)
```
Comment (sreimers) [inline src/httpauth/basic.c:155]: Doxygen hval param is missing

File: src/httpauth/basic.c [hunk 2]
```diff
@@ -123,3 +123,120 @@ int httpauth_basic_encode(const struct httpauth_basic *basic, struct mbuf *mb)
...
+}
+
+
+int httpauth_basic_request_print(struct re_printf *pf,
+	const struct httpauth_basic_req *req)
+{
+	if (!req)
+		return 0;
+
+	if (req->charset && str_len(req->charset))
```
Comment (alfredh) [inline src/httpauth/basic.c:144]: it is better to use "str_isset" here

File: src/httpauth/basic.c [hunk 3]
```diff
@@ -123,3 +123,120 @@ int httpauth_basic_encode(const struct httpauth_basic *basic, struct mbuf *mb)
...
+		return re_hprintf(pf, "Basic realm=\"%s\", charset=\"%s\"",
+			req->realm, req->charset);
+	else
+		return re_hprintf(pf, "Basic realm=\"%s\"", req->realm);
```
Comment (alfredh) [inline src/httpauth/basic.c:148]: this code can be simplified.

Print "basic realm" every time. If charset is set, append charset.

File: src/httpauth/basic.c [hunk 4]
```diff
@@ -123,3 +123,120 @@ int httpauth_basic_encode(const struct httpauth_basic *basic, struct mbuf *mb)
...
+		err = ENOMEM;
+		goto out;
+	}
+
+	err = base64_decode(b64c.p, b64c.l, (uint8_t *) c, &clen);
+	if (err)
+		goto out;
+
+	err = mbuf_printf(mb, "%b:%b",
+		user, str_len(user), passwd, str_len(passwd));
```
Comment (alfredh) [inline src/httpauth/basic.c:195]: can also use %s:%s here ...

File: src/httpauth/basic.c [hunk 5]
```diff
@@ -123,3 +123,120 @@ int httpauth_basic_encode(const struct httpauth_basic *basic, struct mbuf *mb)
...
+	if (err)
+		goto out;
+
+	if (memcmp(mb->buf, c, clen) != 0)
```
Comment (alfredh) [inline src/httpauth/basic.c:199]: it is more secure to use this function:

int mem_seccmp(const uint8_t *s1, const uint8_t *s2, size_t n);

**Comment by alfredh** [comment]
looks good. Please merge to main...

**Pull request merged**